tests/unit/Fun/LazyTest.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Tests\Unit\Fun;
+
+use Generator;
+use PHPUnit\Framework\TestCase;
+use Psl\Fun;
+use RuntimeException;
+
+final class LazyTest extends TestCase
+{
+    public function testItDoesNotLoadInitializerDuringCreation(): void
+    {
+        $this->expectNotToPerformAssertions();
+        Fun\lazy(static function () {
+            throw new RuntimeException('nonono');
+        });
+    }
+
+    public function testItCanBeUsedAsALazyProxy(): void
+    {
+        $proxy = Fun\lazy(static fn (): int => 132);
+
+        static::assertSame(132, $proxy());
+    }
+
+    public function testItCanBeUsedAsALazyProxyAndChainCalls(): void
+    {
+        $x = new class {
+            public function doSomething(): int
+            {
+                return 132;
+            }
+        };
+        $proxy = Fun\lazy(static fn (): object => $x);
+
+        static::assertSame(132, $proxy()->doSomething());
+    }
+
+    public function testItCanBeUsedAsALazyEvaluator(): void
+    {
+        $a = Fun\lazy(static fn(): int => 1);
+        for ($i = 1; $i <= 10; $i++) {
+            $b = $a;
+            $a = Fun\lazy(static fn() => $b() + $b());
+        }
+
+        static::assertSame($a(), pow(2, 10));
+    }
+
+    public function testItCanBeUsedAsALazyStream(): void
+    {
+        $incrementalNumbersStream = Fun\lazy(static function () {
+            $i = 0;
+            while (true) {
+                yield ++$i;
+            }
+        });
+
+        $take = static function (Generator $stream, int $n) {
```

Comment:
FYI: `Dict\take()` and `Dict\take_while()` both try to rewind the closed stream().
This might be something that needs fixing?

Reply:
yea, i guess we should fix these 👍 ( can you open a separate issue? )